MiniDSP.cpp

```diff
@@ -0,0 +1,117 @@
+/* Copyright (C) 2014 Kristian Lauszus, TKJ Electronics. All rights reserved.
+ This software may be distributed and modified under the terms of the GNU
+ General Public License version 2 (GPL2) as published by the Free Software
+ Foundation and appearing in the file GPL2.TXT included in the packaging of
+ this file. Please note that GPL2 Section 2[b] requires that all works based
+ on this software must also be made publicly available under the terms of
+ the GPL2 ("Copyleft").
+ Contact information
+ -------------------
+ Kristian Lauszus, TKJ Electronics
+ Web      :  http://www.tkjelectronics.com
+ e-mail   :  kristianl@tkjelectronics.com
+ */
+
+#include "MiniDSP.h"
+
+namespace {
+uint8_t RequestStatusOutputCommand[] = {0x05, 0xFF, 0xDA, 0x02};
+uint8_t StatusInputCommand[]{0x05, 0xFF, 0xDA};
+
+// Returns first byte of the sum of given bytes.
+uint8_t Checksum(const uint8_t *data, uint16_t nbytes) {
+  int sum = 0;
+  for (int i = 0; i < nbytes; i++) {
+    sum += data[i];
+  }
+
+  return sum & 0xFF;
+}
+} // namespace
+
+void MiniDSP::ParseHIDData(USBHID *hid __attribute__((unused)),
+                           bool is_rpt_id __attribute__((unused)), uint8_t len,
+                           uint8_t *buf) {
+
+  // Only care about valid data for the MiniDSP 2x4HD.
+  if (HIDUniversal::VID != MINIDSP_VID || HIDUniversal::PID != MINIDSP_PID ||
+      len <= 2 || buf == nullptr) {
```

Comment:
This should be `len < 4` or you will have a buffer over-read if `len` is 3, as `StatusInputCommand` is 3 bytes long i.e. you will be reading parsed the `buf`.